src/Polly.Extensions/Telemetry/ResilienceTelemetryDiagnosticSource.cs

```diff
@@ -0,0 +1,123 @@
+using System.Diagnostics.Metrics;
+using System.Globalization;
+using Microsoft.Extensions.Logging;
+using Polly.Strategy;
+using Polly.Telemetry;
+
+namespace Polly.Extensions.Telemetry;
+
+internal class ResilienceTelemetryDiagnosticSource : DiagnosticSource
+{
+    private const string ResilienceEventFormatString =
+        "Resilience event occurred. " +
+        "EventName: '{EventName}', " +
+        "Builder Name: '{BuilderName}', " +
+        "Strategy Name: '{StrategyName}', " +
+        "Strategy Type: '{StrategyType}', " +
+        "Strategy Key: '{StrategyKey}', " +
+        "Outcome: '{Outcome}'";
+
+    private static readonly Meter Meter = new(TelemetryUtil.PollyDiagnosticSource, "1.0");
+
+    private static readonly Action<ILogger, string, string, string, string, string, string, Exception?> ResilienceEventLog = LoggerMessage.Define<string, string, string, string, string, string>(
+            LogLevel.Warning,
+            new EventId(1, "ResilienceEvent"),
+            ResilienceEventFormatString);
+
+    private readonly ILogger _logger;
+    private readonly Counter<int> _counter;
+
+    public ResilienceTelemetryDiagnosticSource(ResilienceStrategyTelemetryOptions options)
+    {
+        LoggerFactory = options.LoggerFactory;
+        Enrichers = options.Enrichers.ToList();
+        OutcomeFormatter = options.OutcomeFormatter;
+
+        _logger = options.LoggerFactory.CreateLogger(TelemetryUtil.PollyDiagnosticSource);
+        _counter = Meter.CreateCounter<int>(
+            "resilience-events",
+            description: "Tracks the number of important resilience events that occurred in resilience strategies.");
+    }
+
+    public ILoggerFactory LoggerFactory { get; }
+
+    public Func<Outcome, string> OutcomeFormatter { get; }
+
+    public List<Action<EnrichmentContext>> Enrichers { get; }
+
+    public override bool IsEnabled(string name) => true;
+
+    public override void Write(string name, object? value)
+    {
+        if (value is not TelemetryEventArguments args)
+        {
+            return;
+        }
+
+        LogEvent(args);
+        MeterEvent(args);
+    }
+
+    private void MeterEvent(TelemetryEventArguments args)
+    {
+        var context = EnrichmentContext.Get(args.Arguments, args.Outcome);
+
+        var tags = new TagList
+        {
+            { "event-name", args.EventName },
+            { "builder-name", args.Source.BuilderName },
+            { "strategy-name", args.Source.StrategyName },
+            { "strategy-type", args.Source.StrategyType },
+            { "result-type", args.Context.IsVoid ? "void" : args.Context.ResultType.Name.ToString(CultureInfo.InvariantCulture) }
+        };
```

Comment:
If this is dictionary-like can we use the new `["key"] = value` syntax instead?

Reply:
It's not a dictionary. Just flat list.